### PR TITLE
[in_app_purchase] Added userIds to google play purchase

### PR DESCRIPTION
**Pull request by renefloor**
This PR adds the obfuscatedAccountId and obfuscatedProfileId to the PurchaseWrapper of a Google Play Purchase.
It contains documentation about how these id's can be set on the purchase.

Fixes #83651
## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format])
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[plugin_tool format]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java [hunk 1]
```diff
@@ -63,6 +64,11 @@
     info.put("developerPayload", purchase.getDeveloperPayload());
     info.put("isAcknowledged", purchase.isAcknowledged());
     info.put("purchaseState", purchase.getPurchaseState());
+    AccountIdentifiers accountIdentifiers = purchase.getAccountIdentifiers();
+    if (accountIdentifiers != null) {
```
Comment (mvanbeusekom) [inline packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java:68]: There is already a test checking that the `obfuscatedAccountId` and `obfuscatedProfileId` are present in the `HashMap` when the the `accountIdentifiers` object is not `null`. However I think it is also good to test that the keys are not present when the `accountIdentifiers` object is `null`.

This guards against somebody removing the `if` statement without thinking about it.

**Review by mvanbeusekom — COMMENTED** [review]
Looks very good, I think the code would benefit from one more test.

File: packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java (hunk 1 above)
Comment (renefloor) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java:68]: @mvanbeusekom this is difficult to test, because although getAccountIdentifiers is nullable, I don't think it will ever be null.

This is the decompiled source code of `getAccountIdentifiers()`
```
    @Nullable
    public AccountIdentifiers getAccountIdentifiers() {
        String var1 = this.zzc.optString("obfuscatedAccountId");
        String var3 = this.zzc.optString("obfuscatedProfileId");
        if (var1 == null && var3 == null) {
            return null;
        } else {
            AccountIdentifiers var2 = new AccountIdentifiers(var1, var3);
            return var2;
        }
    }
```

The account and profile id are retrieved from the JSONObject using [optString(key)](https://developer.android.com/reference/org/json/JSONObject?authuser=1#optString(java.lang.String)).

> Returns the value mapped by name if it exists, coercing it if necessary, or the empty string if no such mapping exists.

Not setting the `obfuscatedAccountId` and `obfuscatedProfileId` in the json will lead to getting a `new AccountIdentifiers("", "")`. Setting them to `null` leads to the same result as that is ignored by the JSONObject decoder. It is marked as `@Nullable` so we can't remove `if (accountIdentifiers != null)`, but it is actually never `null` in the current implementation.

Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase_android/android/src/main/java/io/flutter/plugins/inapppurchase/Translator.java:68]: What if you pass in a mock instance of the `Purchase` class? Something like:

```java
class FakePurchase extends Purchase {
  @Override
  public AccountIdentifiers getAccountIdentifiers() {
    return null;
  }
}
```

**Review by mvanbeusekom — APPROVED** [review]
LGTM
